### ISSUE TITLE
feat(interpolate): kNN / IDW interpolation (idw_to_grid, idw_to_points, fillnan_idw)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/xrtoolz#177]
# Design: kNN / IDW interpolation

**Status:** design draft (MVP)
**Author:** @jejjohnson
**Date:** 2026-05-07 (updated 2026-05-09: switched backend scipy→sklearn)
**Backend:** `sklearn.neighbors.BallTree` / `KDTree` (already a hard dep — see [`pyproject.toml`](../pyproject.toml) `scikit-learn>=1.4`)
**Module home:** new file [`src/xr_toolz/interpolate/_src/knn.py`](../src/xr_toolz/interpolate/_src/knn.py) + operators in [`interpolate/operators.py`](../src/xr_toolz/interpolate/operators.py)
**Related:** [#159](https://github.com/jejjohnson/xr_toolz/issues/159) (grid_to_points), [#161](https://github.com/jejjohnson/xr_toolz/issues/161) (biharmonic inpaint), [#158](https://github.com/jejjohnson/xr_toolz/issues/158) (dask), [#181](https://github.com/jejjohnson/xr_toolz/issues/181) (KDE points-to-grid — sibling sklearn-backed primitive)

---

## 1. User story

> As a user with **scattered observations** (Argo profile locations, drifter positions, AIS, weather stations), I want to interpolate them **onto a regular grid** (or onto another set of points, or fill missing pixels in a gridded field) using **inverse-distance weighting** — the standard fast, transparent, hyperparameter-light technique.

The existing `points_to_grid` aggregates by binned statistics (mean / sum per cell), which throws information away when a cell has zero hits and ignores neighbours when a cell has many. kNN/IDW fills the gap.

## 2. Why sklearn.neighbors (and not scipy.spatial.cKDTree)

| Option | Verdict |
|---|---|
| **`sklearn.neighbors.BallTree` / `KDTree`** | **Pick.** scikit-learn is already a hard dep (`pyproject.toml`). Same C-level performance as `cKDTree` for Euclidean queries; **`BallTree` adds `metric="haversine"`** for great-circle kNN — a feature `cKDTree` does not have. Stateless query API (`tree.query(...)`) — no `fit`/`predict` round-trip required. |
| `scipy.spatial.cKDTree` | Fast (C++), simple, but Euclidean-only — forces (lon, lat) pole bias on global fields. No haversine. |
| `sklearn.neighbors.KNeighborsRegressor` | Adds `fit`/`predict` machinery for what is fundamentally a stateless query. Heavier than needed; the underlying tree is what we actually want. |
| Hand-rolled brute force | Fine for small N, but kNN benefits from a tree above ~10⁴ points. |

**Geoscience-specific wins from sklearn:**

1. **Spherical kNN.** `BallTree(metric="haversine")` gives great-circle nearest neighbours directly — eliminates the pole-bias caveat that sat in §10 of the previous design. Most-requested feature for global scattered-obs interpolation.
2. **Tree-based scaling.** `KDTree` and `BallTree` are both C-level; performance is comparable to `cKDTree` for Euclidean. Nothing lost.
3. **Already in our stack.** Adopting sklearn here is consistent with [#181](https://github.com/jejjohnson/xr_toolz/issues/181) (KDE) and [#178](https://github.com/jejjohnson/xr_toolz/issues/178) (GP), which are also sklearn-backed.

For the future GP/kriging primitive ([#178](https://github.com/jejjohnson/xr_toolz/issues/178)) we use `sklearn.gaussian_process` — same stack.

## 3. Scope (MVP)

In:
- **`idw_to_grid`** — scattered (lon, lat, value) → regular grid via inverse-distance weighting.
- **`idw_to_points`** — scattered → arbitrary target points.
- **`fillnan_idw`** — fill NaNs in a gridded field by IDW from finite neighbours (drop-in alternative to `griddata` / `RBFInterpolator`).
- **`metric: Literal["euclidean", "haversine"] = "euclidean"`** on all three functions. `"haversine"` evaluates great-circle distances on the sphere (inputs converted to radians internally; `max_distance` interpreted as radians).
- **`IDWToGrid`**, **`IDWToPoints`**, **`FillNaNIDW`** operators.

Out (file as follow-ups):
- **Anisotropic distance metrics** (e.g. zonal/meridional weighting beyond what BallTree's standard metrics support).
- **Adaptive `k`** based on local point density.
- **Gaussian Process / kriging** — separate issue (uncertainty-aware, smoothness, expensive). [#178](https://github.com/jejjohnson/xr_toolz/issues/178)

## 4. Mathematics

For target point $\mathbf{x}^*$ and source points $\{(\mathbf{x}_i, v_i)\}$, find the $k$ nearest neighbours and compute

$$
\hat v(\mathbf{x}^*) = \frac{\sum_{i \in N_k(\mathbf{x}^*)} w_i\, v_i}{\sum_{i \in N_k(\mathbf{x}^*)} w_i},
\qquad
w_i = \frac{1}{(d_i + \epsilon)^p},
$$

with $d_i = \|\mathbf{x}^* - \mathbf{x}_i\|$ in the chosen metric (Euclidean in (lon, lat) or haversine on the sphere), power exponent $p$ (default 2 — physical inverse-square weighting), and a small $\epsilon$ to avoid divide-by-zero when a target coincides exactly with a source point.

**Edge cases:**
- Exact coincidence ($d_i \to 0$): return $v_i$ exactly (use `eps = 0` plus a check).
- Zero finite neighbours within `max_distance` (if specified): return NaN.
- Fewer than `k` finite neighbours total: use what's available, return NaN if none.

### 5.1 Layer 0 — `_src/knn.py`

```python
def idw_to_grid(
    lons: ArrayLike, lats: ArrayLike, values: ArrayLike,
    grid: Grid,
    *,
    k: int = 8,
    power: float = 2.0,
    metric: Literal["euclidean", "haversine"] = "euclidean",
    max_distance: float | None = None,
    eps: float = 1e-12,
) -> xr.DataArray:
    """Inverse-distance interpolate scattered points onto ``grid``.

    For ``metric="haversine"``, ``max_distance`` is interpreted in radians
    (e.g. ``50 / 6371`` for ~50 km).
    """

def idw_to_points(
    src_lons, src_lats, src_values,
    dst_lons, dst_lats,
    *, k: int = 8, power: float = 2.0,
    metric: Literal["euclidean", "haversine"] = "euclidean",
    max_distance: float | None = None,
    eps: float = 1e-12,
) -> np.ndarray: ...

def fillnan_idw(
    da: xr.DataArray, *,
    lon: str = "lon", lat: str = "lat",
    k: int = 8, power: float = 2.0,
    metric: Literal["euclidean", "haversine"] = "euclidean",
    max_distance: float | None = None,
) -> xr.DataArray:
    """Fill NaNs in a 2-D lon/lat field by IDW from the finite neighbours.

    For ``metric="haversine"`` we use the (lat, lon) coordinates as
    great-circle inputs; for ``"euclidean"`` we use grid-index distances
    (column / row) for backwards-compat with earlier behaviour.
    """
```

### 5.2 Layer 1 — `interpolate/operators.py`

```python
class IDWToGrid(Operator):
    """Wrap :func:`idw_to_grid`. Input: ``(lons, lats, values)`` 3-tuple
    (mirrors :class:`PointsToGrid`)."""

class IDWToPoints(Operator):
    """Wrap :func:`idw_to_points`. Configured with target points; input
    is the source ``(lons, lats, values)`` tuple."""

class FillNaNIDW(Operator):
    """Wrap :func:`fillnan_idw`."""
```

## 6. Implementation sketch

```python
# _src/knn.py
import numpy as np
import xarray as xr
from sklearn.neighbors import BallTree, KDTree


def _build_tree(src_xy: np.ndarray, metric: str):
    # src_xy: (N, 2) in (lon, lat) deg if euclidean, (lat, lon) rad if haversine.
    if metric == "haversine":
        return BallTree(src_xy, metric="haversine")
    if metric == "euclidean":
        # KDTree is the natural choice for Euclidean and matches cKDTree perf.
        return KDTree(src_xy, metric="euclidean")
    raise ValueError(f"unknown metric {metric!r}")


def _to_metric_xy(lons: np.ndarray, lats: np.ndarray, metric: str) -> np.ndarray:
    if metric == "haversine":
        # BallTree wants (lat, lon) in radians.
        return np.deg2rad(np.column_stack([lats, lons]))
    return np.column_stack([lons, lats])  # euclidean: (lon, lat) deg


def _idw_kernel(tree, src_values, queries, k, power, max_distance, eps):
    # queries: (M, 2) in tree-native units. Returns (M,).
    k_eff = min(k, len(src_values))
    d, idx = tree.query(queries, k=k_eff)
    if k_eff == 1:
        d, idx = d[:, None], idx[:, None]
    # Out-of-range mask (sklearn returns finite distances; we mask post-hoc).
    if max_distance is not None:
        valid = d <= max_distance
    else:
        valid = np.ones_like(d, dtype=bool)
    exact = (d == 0) & valid
    weights = np.where(valid, 1.0 / (d + eps) ** power, 0.0)
    vals = src_values[idx]
    num = (weights * vals).sum(axis=1)
    den = weights.sum(axis=1)
    out = np.where(den > 0, num / den, np.nan)
    # Exact-hit shortcut: return that source value verbatim.
    if exact.any():
        first_exact = exact.argmax(axis=1)
        out = np.where(
            exact.any(axis=1),
            vals[np.arange(len(vals)), first_exact],
            out,
        )
    return out


def idw_to_grid(
    lons, lats, values, grid, *,
    k=8, power=2.0, metric="euclidean",
    max_distance=None, eps=1e-12,
):
    lons = np.ravel(np.asarray(lons))
    lats = np.ravel(np.asarray(lats))
    v = np.ravel(np.asarray(values, dtype=np.float64))
    finite = np.isfinite(v) & np.isfinite(lons) & np.isfinite(lats)
    if not finite.any():
        raise ValueError("idw_to_grid requires at least 1 finite source")
    src_xy = _to_metric_xy(lons[finite], lats[finite], metric)
    tree = _build_tree(src_xy, metric)

    lon_g, lat_g = np.meshgrid(grid.lon, grid.lat, indexing="xy")
    queries = _to_metric_xy(lon_g.ravel(), lat_g.ravel(), metric)
    out = _idw_kernel(
        tree, v[finite], queries, k, power, max_distance, eps,
    )
    return xr.DataArray(
        out.reshape(len(grid.lat), len(grid.lon)),
        dims=("lat", "lon"),
        coords={"lat": grid.lat, "lon": grid.lon},
    )


def fillnan_idw(
    da, *, lon="lon", lat="lat",
    k=8, power=2.0, metric="euclidean", max_distance=None,
):
    lon_coord = np.asarray(da.coords[lon].values)
    lat_coord = np.asarray(da.coords[lat].values)

    def _fill_slice(arr2d):
        finite = np.isfinite(arr2d)
        if finite.all() or not finite.any():
            return arr2d
        if metric == "haversine":
            lon_g, lat_g = np.meshgrid(lon_coord, lat_coord, indexing="xy")
            src_xy = _to_metric_xy(
                lon_g[finite], lat_g[finite], "haversine",
            )
            qry_xy = _to_metric_xy(
                lon_g[~finite], lat_g[~finite], "haversine",
            )
            tree = _build_tree(src_xy, "haversine")
        else:
            yy, xx = np.indices(arr2d.shape)
            src_xy = np.column_stack([xx[finite], yy[finite]])
            qry_xy = np.column_stack([xx[~finite], yy[~finite]])
            tree = _build_tree(src_xy, "euclidean")
        filled = _idw_kernel(
            tree, arr2d[finite].astype(np.float64),
            qry_xy, k, power, max_distance, 1e-12,
        )
        out = arr2d.copy()
        out[~finite] = filled
        return out

    return xr.apply_ufunc(
        _fill_slice, da,
        input_core_dims=[[lat, lon]], output_core_dims=[[lat, lon]],
        vectorize=True, dask="parallelized",
        output_dtypes=[da.dtype],
        dask_gufunc_kwargs={"allow_rechunk": False},
    )
```

## 7. Usage examples

```python
from xr_toolz.interpolate import idw_to_grid, fillnan_idw, FillNaNIDW
from xr_toolz.interpolate._src.binning import Grid

# Argo SST onto a regular regional grid (Euclidean is fine here)
grid = Grid(lon=np.arange(-80, -10, 0.25), lat=np.arange(20, 60, 0.25))
sst_grid = idw_to_grid(lons, lats, sst_values, grid, k=8, power=2.0)

# Global drifter T/S onto a 1° grid — haversine for correct distances near poles
grid = Grid(lon=np.arange(-180, 180, 1.0), lat=np.arange(-80, 80, 1.0))
sst_grid = idw_to_grid(
    lons, lats, sst_values, grid,
    k=8, power=2.0,
    metric="haversine",
    max_distance=200.0 / 6371.0,  # 200 km radius in radians
)

# Cloud-gap fill in a gridded SST field (uses index distance by default)
sst_filled = fillnan_idw(sst, k=8, power=2.0)
```

## 8. Tests (DoD)

1. **Identity at source nodes:** target == one source point → exact value.
2. **Constant field:** $v_i = c$ → output uniformly $c$.
3. **Power-law sanity:** higher `power` → output closer to nearest neighbour; `power=0` → unweighted mean of `k` neighbours.
4. **`max_distance` honored:** queries farther than `max_distance` from any source return NaN. Test in both Euclidean (degrees) and haversine (radians) units.
5. **Haversine vs Euclidean.** Two source points 180° apart on the equator vs near the pole: haversine gives consistent neighbour distances; Euclidean does not.
6. **`fillnan_idw` round-trip:** drill a hole in a smooth field, fill, RMSE bounded.
7. **Comparison guard:** on a fixed synthetic, `fillnan_idw` RMSE in the gap is *comparable to* `griddata("linear")` (locks in the "this is a useful alternative" claim — IDW is faster but slightly less accurate).
8. **Dask compat:** chunked along leading dims; chunking-contract error if chunked along `lat`/`lon`.
9. **Operator round-trip.**

## 9. Subtasks

1. Implement kernel + 3 functions in `_src/knn.py` (sklearn-backed).
2. Add operators in `interpolate/operators.py`.
3. Public exports.
4. Tests per §8 — including the haversine-vs-Euclidean global-vs-pole test.
5. Docs notebook: scattered Argo SST → IDW grid; cloud-gap fill side-by-side with biharmonic / griddata. Include a global-drifter haversine example.

## 10. Open questions

- **Default `k`?** 8 is a common heuristic for 2-D scattered data. Too small (k=4) under-smooths, too large (k=20) over-smooths. **Recommendation:** `k=8`. Document and let users tune.
- **Default `power`?** 2 is the physical default (inverse-square — Shepard's original 1968 method). Others use 1 (gentler) or 3+ (more local). **Recommendation:** `power=2`.
- **Default `metric`.** Euclidean preserves backwards-compat with the most common (regional, projected) use case and matches user intuition for (lon, lat) grids; haversine is opt-in. **Recommendation:** `"euclidean"`, document haversine as the right call for global / near-polar work.
- **`max_distance` units.** Degrees (Euclidean) vs radians (haversine) is a footgun. **Document inline** with a "≈ km / 6371" rule of thumb in the docstring; raise no exception (sklearn handles both correctly given matching units).
- **Tree choice.** `KDTree` for Euclidean is essentially identical to `cKDTree` in our regime; `BallTree` is required for haversine. **Recommendation:** auto-select per metric (transparent to caller).
- **Should `fillnan_idw` go in `gap_fill.py` instead of `knn.py`?** Symmetry argument: the other fillers are there. **Recommendation:** keep the kernel in `knn.py` but also export from `gap_fill.py` namespace for discovery.
- **`k=1` (nearest)** as a special case. Equivalent to `griddata(method="nearest")`. **Recommendation:** support but suggest `griddata` for that use case.
- **Performance ceiling.** sklearn's tree query is fast (C-level) and tree construction is O(N log N). At >10⁶ points consider chunking the query (the tree itself fits fine). **Punt** until a user hits it.
- **Cross-validated `k` / `power`.** Trivial to add via `GridSearchCV(KNeighborsRegressor(weights="distance"), ...)` since we're on sklearn — but only useful when users have held-out data. **Punt** to follow-up.
